### ISSUE TITLE
Accented characters are always displayed in lowercase

### ISSUE THREAD

**Issue by Jaffacakelover** [RetroPie/EmulationStation#724]
The content scraper (and any rom lists for systems that have been scraped) displays accented characters in lower case, while everything else is upper case.
Most noticeable with Pokémon games. For example:
![IMG_20201031_221610_cropped](https://user-images.githubusercontent.com/3490850/104201078-43053200-5421-11eb-884a-d535897a32c9.jpg)

The font that RetroPie uses (Open Sans Condensed) does include upper case accented characters: For example, `É` is `U+00C9` in unicode.

**Comment by Gemba** [comment]
As you are right, @Jaffacakelover the font does provide this character (and also the "Cabin Bold" (which is used in the game list for Carbon Theme).

However, root cause is that how the uppercase function is implemented in ES. 
The i18n fork of ES [uses Boost Lib](https://github.com/flyinghead/EmulationStation/commit/588f81f1809badf3d706e4367393e10711bc464c) to tackle this, but this ES does not use Boost any longer. So I propose a "boost-less" change:

```cpp
#include <iostream>
#include <locale>
#include <codecvt>

// yarked from current master:
// https://github.com/RetroPie/EmulationStation/blob/771f2a608b84b5262666af7888f377a5b3f280ea/es-core/src/utils/StringUtil.cpp#L163-L172
std::string toUpper(const std::string& _string)
{
        std::string string;

        for(size_t i = 0; i < _string.length(); ++i)
                string += (char)toupper(_string[i]);

        return string;

}

// proposed change/fix for #724
std::string toUpperNew(const std::string& _string)
{
        std::wstring w_str = std::wstring_convert<std::codecvt_utf8<wchar_t> >().from_bytes(_string);
        auto& fct = std::use_facet<std::ctype<wchar_t> >(std::locale());
        fct.toupper(&w_str[0], &w_str[0] + w_str.size());
        return std::wstring_convert<std::codecvt_utf8<wchar_t> >().to_bytes(w_str);
}

int main() {
        std::locale::global(std::locale(""));

        std::string str = "Déjà Vu: A Nightmare Come True";

        std::cout << "\nInput string    : " << str << std::endl;
        std::cout << "\nUpper (current) : " << toUpper(str) << std::endl;
        std::cout << "\nUpper (proposed): " << toUpperNew(str) << std::endl;

        // enable for profiling
        /*
        std::string tmp;
        for (int i = 0 ; i < 20 * 1000; i++) {
            //tmp = toUpper(str);
            tmp = toUpperNew(str);
        }
        */

        return 0;
}
```
(used: `g++-8 uppercase_wstring.cpp -o uppercase_wstring`)

Will output with locale `en_GB.UTF-8` on RetroPie:
```
./uppercase_wstring

Input string    : Déjà Vu: A Nightmare Come True

Upper (current) : DéJà VU: A NIGHTMARE COME TRUE

Upper (proposed): DÉJÀ VU: A NIGHTMARE COME TRUE
```

Verified ok on Linux/recent RetroPie 4.7.1 (the impact on runtime can be neglected IMHO).
Could someone (of the core maintainers, e.g. @pjft @tomaz82 ...) pls verify this proposed change on Windows/Mac before proceeding?

**Comment by cmitu** [comment]
Doesn't work on macOS (Apple clang 11):
``` clang
es-core/src/utils/StringUtil.cpp:165:43: error: no member named 'ctype' in namespace 'std'; did you mean 'wctype'?
                        auto& fct = std::use_facet<std::ctype<wchar_t> >(std::locale());
                                                   ~~~~~^~~~~
                                                        wctype
``` 
According to [cppreference](https://en.cppreference.com/w/cpp/locale/wstring_convert), `std::wstring_convert` is deprecated since C++17.

**Comment by Gemba** [comment]
Ah, thanks. If/if not MacOS is supported by this repo escaped my mind.
I have seen the deprecation note and also read it shall not be "cut off" until some more suitable approach can be provided. 

However, it boils down to a design decision: leave it as it, use this approach (albeit deprecated), use a library (which, if not boost?) or roll your own.

**Comment by Gemba** [comment]
Could verify successfully on Windows too (W10/VS2019). Works with same output and with `std::locale::global(std::locale(""));` iff UTF-8 is enabled in Windows: Open _Region_ -> Register _Administrative_ -> _Change system locale..._ -> Enable Checkbox _Use Unicode UTF-8 ..._

**Comment by tomaz82** [comment]
Only done some quick test, and as mentioned, this requires locale to be changed from "C" to "", which breaks other things, (most noticeable is that Carbon theme layout broke) but I have not had time to investigate how or why it broke.

So if the choice is between broken themes and lowercase accented e, I know which one I prefer.

**Comment by Gemba** [comment]
No need to make a choice. 
Here's what I did: Pulled latest ES (771f2a6)  and latest [ES-carbon-Theme](https://github.com/RetroPie/es-theme-carbon) (b09973e). Applied the proposed changes (including the region setting) from above and let cmake do the tedious work. Voilà!

Works for me:

![ES_2 10 0RP-DEV](https://user-images.githubusercontent.com/418292/107434672-593f0480-6b2b-11eb-9c7f-b54cb2e23131.png)

**Comment by tomaz82** [comment]
Yeah that's not the latest VIP Carbon, and that's beside the point, if this breaks the current VIP Carbon, it might break many other themes, which is why this needs to be investigated further, I don't remember the details anymore but I know I did a lot of testing back when I removed boost before coming to the conclusion that locale( "C" ) was the one that behaved as it should.

**Comment by Gemba** [comment]
Now we are getting closer to the facts. If you let me know where this "VIP Carbon" can be found and also can provide specific details what broke in this "VIP Carbon" we might be able to tackle this issue. Or maybe you table it on the RetroPie forum and ask for testers (for Windows builds).

**Comment by tomaz82** [comment]
I just tested it on "normal" Carbon and it flatout crashes with locale( "" ), and now I remember why I chose to use locale( "C" ) when I removed boost.
locale( "" ) uses the default locale of the computer it's running on, so for example, Swedish, English, Russian, Japanese locales doesn't behave the same.

Swedish for example does not use numbers with periods, but with comma, so it doesn't handle xml values like 0.30, since it expects it to be 0,30, so when it sees 0.30 it just becomes 0. In this case it meant that fontSize became 0 which is invalid. I'm sure there might be other ways around this, but this is where we are right now.

**Comment by tomaz82** [comment]
I also did some benchmarking

```
Function    NumCalls    Time
toUpper        142095     0.062938
toUpperNew     142095     7.365905
```

**Comment by tomaz82** [comment]
This is gonna be a long comment, but I have now done quite a bit of testing. I wanna point out that all my tests have been done on Windows 10 so the results might differ ona pi.

First off, the local that in my tests works the best is:
`std::locale::global(std::locale("en_US.UTF-8"));`

Simply because this gives us the ability to do upper case é without breaking the paring of floats from xml files.

And onto the toUpper itself I made 4 versions and profiled them (PR https://github.com/RetroPie/EmulationStation/pull/716 adds a profiling utility)

But first I added 2 new functions to StringUtil which will be used later on:
```
		std::wstring multiByteToWideChar(const std::string& _string)
		{
			std::wstring string;

			string.resize(_string.length());
			mbstowcs((wchar_t*)string.c_str(), _string.c_str(), string.length());

			return string;

		} // multiByteToWideChar

		std::string wideCharToMultiByte(const std::wstring& _string)
		{
			std::string string;

			string.resize(_string.length());
			wcstombs((char*)string.c_str(), _string.c_str(), string.length());

			return string;

		} // wideCharToMultiByte
```
Then I added these 4 versions of toUpper:
```
		std::string toUpperStd(const std::string& _string)
		{
			ProfileScope("toUpperStd");

			std::wstring w_str = std::wstring_convert<std::codecvt_utf8<wchar_t> >().from_bytes(_string);
			auto& fct = std::use_facet<std::ctype<wchar_t> >(std::locale());
			fct.toupper(&w_str[0], &w_str[0] + w_str.size());
			std::string string = std::wstring_convert<std::codecvt_utf8<wchar_t> >().to_bytes(w_str);

			return string;

		} // toUpperStd

		std::string toUpperC2U(const std::string& _string)
		{
			ProfileScope("toUpperC2U");

			const std::ctype<wchar_t>& facet = std::use_facet<std::ctype<wchar_t>>(std::locale());
			std::wstring               wstring;
			std::string                string;

			size_t i = 0;
			while(i < _string.length())
				wstring += facet.toupper((wchar_t)Utils::String::chars2Unicode(_string, i));

			for(size_t i = 0; i < wstring.length(); ++i)
				string += unicode2Chars(wstring[i]);

			return string;

		} // toUpperC2U

		std::string toUpperMB2WC(const std::string& _string)
		{
			ProfileScope("toUpperMB2WC");

			const std::ctype<wchar_t>& facet   = std::use_facet<std::ctype<wchar_t>>(std::locale());
			std::wstring               wstring = multiByteToWideChar(_string);

			for(size_t i = 0; i < wstring.length(); ++i)
				wstring[i] = facet.toupper(wstring[i]);

			return wideCharToMultiByte(wstring);

		} // toUpperMB2WC

		std::string toUpperMB2WC2(const std::string& _string)
		{
			ProfileScope("toUpperMB2WC2");

			const std::ctype<wchar_t>& facet   = std::use_facet<std::ctype<wchar_t>>(std::locale());
			std::wstring               wstring = multiByteToWideChar(_string);

			facet.toupper(&wstring[0], &wstring[0] + wstring.size());

			return wideCharToMultiByte(wstring);

		} // toUpperMB2WC2
```
A quick explanation of each function:
toUpperStd: The original method by Gemba who initiated this discussion.
toUpperC2U: A different approach using the already existing chars2Unicode and unicode2Chars in for loops.
toUpperMB2WC: Yet a different approach using the 2 new methods multiByteToWideChar and wideCharToMultiByte.
toUpperMB2WC2: A modification of the previous method to avoid the for loop when calling facet.toupper.

I also modfied toUpper to call all 4 of these just so they could be profiled, all 4 variants have been tested and verified to make é to the proper uppercase, everything have also been tested as to not break chinese characters:
```
		std::string toUpper(const std::string& _string)
		{
			toUpperC2U(_string);
			toUpperStd(_string);
			toUpperMB2WC(_string);
			toUpperMB2WC2(_string);

			ProfileScope("toUpper");

			std::string string;

			for(size_t i = 0; i < _string.length(); ++i)
				string += (char)toupper(_string[i]);

			return string;

		} // toUpper
```
None of the results from the 4 new methods in the profiling are actually used as to not favor any of them, and I did put toUpperStd as second one on purpose to show that caching does not come into play here.

The results were as follow on PC:

Debug:
```
Message           Calls    Total Time

toUpper          142307      0.567588
toUpperMB2WC2    142307      0.812284
toUpperMB2WC     142307      1.002347
toUpperC2U       142307      2.892302
toUpperStd       142307     25.876925
```
The winner here is toUpperMB2WC2.

Release:
```
Message           Calls    Total Time

toUpper           142095     0.062217
toUpperMB2WC2     142095     0.102216
toUpperC2U        142095     0.105076
toUpperMB2WC      142095     0.131224
toUpperStd        142095     7.800818
```
The winner is once again toUpperMB2WC2.

What's everyones thought on this?

**Comment by cmitu** [comment]
@tomaz82 nicely done. 
Since `multiByteToWideChar` and `wideCharToMultiByte` and Win32 specific, they won't work on Linux.
I'll try the other 2 functions and see how they compare on a Pi.

**Comment by tomaz82** [comment]
They use mbstowcs and wcstombs which should not be windows specific.

**Comment by cmitu** [comment]
@tomaz82 yes, I forgot to include the implementation and assumed to be the Win32 equivalents. Here's a result from a Pi4 (no Debug/Release configurations):
```
Message             	       Calls	    Total Time

toUpper            	       13579	    0.152930
toUpperC2U         	       13579	    0.023170
toUpperMB2WC       	       13579	    0.024451
toUpperMB2WC2      	       13579	    0.022319
toUpperStd         	       13579	    0.030516
```

**Comment by Gemba** [comment]
@tomaz82 thanks for the throughout analysis and data. If there are better solutions to this which work on Linux _and_ Windows with satisfaction I am open to it. Looks `toUpperMB2WC2` is the way to go, tested on my Rpi system and does not impact runtime.
So only the theme-engine must be instructed to interpret number strings without i18n to be able to form a PR. Right?

**Comment by tomaz82** [comment]
Well no, not really, that thing has been solved by setting
```cpp
std::setlocale(LC_NUMERIC, "C");
```
The problem is what locale the global locale should be set to, it can't be "" because that uses whatever is the default on the running system, which might not even be UTF-8. And that would break even more things.

This is exactly why it was left as "C", 99.9% of the things works, and it's guaranteed to work identically on all systems.

**Comment by Gemba** [comment]
Ah, yes. `LC_NUMERIC` is the straight forward solution (should have had more coffee today morning :wink: ). Plus, if at any time later floats will be implemented in `es_settings.cfg` the values won't be misinterpreted.

RetroPie ships locale `en_GB.UTF-8` by default. 
ES alone on Windows can not rely on this, as you noted correctly, and should test if the current locale is "UTF-8able" and only then switch away from "C" locale.

**Comment by cmitu** [comment]
> The problem is what locale the global locale should be set to, it can't be "" because that uses whatever is the default on the running system, which might not even be UTF-8. And that would break even more things.

Looking at previous changes for locale settings via `std::locale::global`:
- it was originally added in https://github.com/Aloshi/EmulationStation/commit/ac37765744c73bed92c1a7af26e5d0ddc0734148 (_ Use user locale at startup. Should fix Unicode paths on Windows._) :
   ``` cpp
   std::locale::global(boost::locale::generator().generate(""));
   boost::filesystem::path::imbue(std::locale());
   ```
 Looks like the locale was set to the user's runtime locale.

- @tomaz82 refactored the code in #276:
  ``` diff
  -   std::locale::global(boost::locale::generator().generate(""));
  +   std::locale::global(std::locale(std::locale(""), "C", std::locale::numeric));
  ```
  Looks like the user locale was still used, but the `LC_NUMERIC` facet was set to `C`, to ensure the theme units were un-ambiguously intepreted. This however seems to have broken something else, which lead to the next change.

- the latest change was PR #294 which replaced the user's runtime locale with `C` overall:
 
 ``` diff
-   std::locale::global(std::locale(std::locale(""), "C", std::locale::numeric));
+   std::locale::global(std::locale("C"));
  ```

 I cannot find a reference of the issue that triggered the latest change (on Github), but testing the previous version of the code with Japanese characters does indeed shows garbled text (_mojibake_) when the game title is in Japanese (UTF8) **and** the theme has `forceUppercase` set. Didn't experience a crash though.

 Looks like one way or the other there's some breakage in displaying text. Most likely using something like [ICU](https://icu.unicode.org/) would solve the issue without relying on the locale support for upcase/lowercase.
